scipy/_lib/doccer.py

```diff
@@ -202,73 +265,106 @@ def indentcount_lines(lines):
     return indentno
 
 
-def filldoc(docdict, unindent_params=True):
-    ''' Return docstring decorator using docdict variable dictionary
+def filldoc(docdict: dict[str, str], unindent_params: bool = True) -> Decorator[R]:
+    """Return docstring decorator using docdict variable dictionary.
 
     Parameters
     ----------
-    docdict : dictionary
-        dictionary containing name, docstring fragment pairs
-    unindent_params : {False, True}, boolean, optional
-        If True, strip common indentation from all parameters in
-        docdict
+    docdict : dict[str, str]
+        A dictionary containing name, docstring fragment pairs.
+    unindent_params : bool, optional
+        If True, strip common indentation from all parameters in docdict.
+        Default is False.
 
     Returns
     -------
     decfunc : function
-        decorator that applies dictionary to input function docstring
-
-    '''
+        The decorator function that applies dictionary to its
+        argument's __doc__ attribute.
+    """
     if unindent_params:
         docdict = unindent_dict(docdict)
 
-    def decorate(f):
-        f.__doc__ = docformat(f.__doc__, docdict)
-        return f
+    def decorate(func: Callable[P, R]) -> Callable[P, R]:
+        # __doc__ may be None for optimized Python (-OO)
+        doc = func.__doc__ or ""
+        func.__doc__ = docformat(doc, docdict)
+        return func
+
     return decorate
 
 
-def unindent_dict(docdict):
-    ''' Unindent all strings in a docdict '''
-    can_dict = {}
+def unindent_dict(docdict: dict[str, str]) -> dict[str, str]:
```

Comment:
See my earlier comment about using `Mapping` here for the parameter.
The `dict[str, str]` return type is perfectly fine 👌🏻